File: lib/gherkin3/pickles/compiler.rb
require_relative '../dialect'

module Gherkin3
  module Pickles
    class Compiler
      def compile(feature, path)
        pickles = []
        dialect = Dialect.for(feature[:language])

        feature_tags = feature[:tags]
        background_steps = get_background_steps(feature[:background])

        feature[:scenarioDefinitions].each do |scenario_definition|
          if(scenario_definition[:type] == :Scenario)
            compile_scenario(feature_tags, background_steps, scenario_definition, dialect, path, pickles)
          else
            compile_scenario_outline(feature_tags, background_steps, scenario_definition, dialect, path, pickles)
          end
        end
        return pickles
      end

    private

      def compile_scenario(feature_tags, background_steps, scenario, dialect, path, pickles)
        steps = [].concat(background_steps)

        tags = [].concat(feature_tags).concat(scenario[:tags])

        scenario[:steps].each do |step|
          steps.push(pickle_step(step))
        end

        pickle = {
          path: path,
          tags: pickle_tags(tags),
          name: scenario[:keyword] + ": " + scenario[:name],
          locations: [pickle_location(scenario[:location])],
          steps: steps
        }
        pickles.push(pickle)
      end

      def compile_scenario_outline(feature_tags, background_steps, scenario_outline, dialect, path, pickles)
        keyword = dialect.scenario_keywords[0]
        scenario_outline[:examples].each do |examples|
          variable_cells = examples[:tableHeader][:cells]
          examples[:tableBody].each do |values|
            value_cells = values[:cells]
            steps = [].concat(background_steps)
            tags = [].concat(feature_tags).concat(scenario_outline[:tags]).concat(examples[:tags])

            scenario_outline[:steps].each do |scenario_outline_step|
              step_text = interpolate(scenario_outline_step[:text], variable_cells, value_cells);
              arguments = create_pickle_arguments(scenario_outline_step[:argument], variable_cells, value_cells)
              pickle_step = {
                text: step_text,
                arguments: arguments,
                locations: [
                  pickle_location(values[:location]),
                  pickle_location(scenario_outline_step[:location])
                ]
              }
              steps.push(pickle_step)
            end

            pickle = {
              path: path,
              name: keyword + ": " + interpolate(scenario_outline[:name], variable_cells, value_cells),
              steps: steps,
              tags: pickle_tags(tags),
              locations: [
                pickle_location(values[:location]),
                pickle_location(scenario_outline[:location])
              ]
            }
            pickles.push(pickle);

          end
        end
      end

      def create_pickle_arguments(argument, variables, values)
        result = []
        return result if argument.nil?
        if (argument[:type] == :DataTable)
          table = {
            rows: argument[:rows].map do |row|
              {
                cells: row[:cells].map do |cell|
                  {
                    location: pickle_location(cell[:location]),
                    value: interpolate(cell[:value], variables, values)
                  }
                end
              }
            end
          }
          result.push(table)
        elsif (argument[:type] == :DocString)
          doc_string = {
            location: pickle_location(argument[:location]),
            content: interpolate(argument[:content], variables, values)
          }
          result.push(doc_string)
        else
          raise 'Internal error'
        end
        result
      end

      def interpolate(name, variable_cells, value_cells)
        variable_cells.each_with_index do |variable_cell, n|
          value_cell = value_cells[n]
          name = name.gsub('<' + variable_cell[:value] + '>', value_cell[:value])
        end
        name
      end

      def get_background_steps(background)
        if(background)
          background[:steps].map do |step|
            pickle_step(step)
          end
        else
          []
        end
      end

      def pickle_step(step)
        {
          text: step[:text],
          arguments: create_pickle_arguments(step[:argument], [], []),
          locations: [pickle_location(step[:location])]
        }
      end

      def pickle_location(location)
        {
          line: location[:line],
          column: location[:column]
        }
      end

      def pickle_tags(tags)
        tags.map {|tag| pickle_tag(tag)}
      end

      def pickle_tag(tag)
        {
          name: tag[:name],
          location: pickle_location(tag[:location])
        }
      end
    end
  end
end

File: python/gherkin3/pickles/compiler.py
import re

from ..dialect import Dialect
from ..count_symbols import count_symbols

def compile(feature, path):
    pickles = []
    dialect = feature['language']
    feature_tags = feature['tags']
    background_steps = _get_background_steps(feature, path)
    for scenario_definition in feature['scenarioDefinitions']:
        args = (feature_tags, background_steps, scenario_definition,
                dialect, path, pickles)
        if scenario_definition['type'] is 'Scenario':
            _compile_scenario(*args)
        else:
            _compile_scenario_outline(*args)
    return pickles


def _compile_scenario(feature_tags, background_steps, scenario,
                      dialect, path, pickles):
    steps = list(background_steps)
    tags = list(feature_tags) + list(scenario['tags'])

    for step in scenario['steps']:
        steps.append(_pickle_step(step, path))

    pickle = {
        'tags': _pickle_tags(tags, path),
        'name': u'{0[keyword]}: {0[name]}'.format(scenario),
        'locations': [_pickle_location(scenario['location'], path)],
        'steps': steps
    }
    pickles.append(pickle)


def _compile_scenario_outline(feature_tags, background_steps, scenario_outline,
                              dialect, path, pickles):
    keyword = Dialect.for_name(dialect).scenario_keywords[0]

    for examples in scenario_outline['examples']:
        variable_cells = examples['tableHeader']['cells']

        for values in examples['tableBody']:
            value_cells = values['cells']
            steps = list(background_steps)
            tags = list(feature_tags) \
                + list(scenario_outline['tags']) \
                + list(examples['tags'])

            for scenario_outline_step in scenario_outline['steps']:
                step_text = _interpolate(
                    scenario_outline_step['text'],
                    variable_cells,
                    value_cells)
                arguments = _create_pickle_arguments(
                    scenario_outline_step.get('argument'),
                    variable_cells,
                    value_cells,
                    path)
                _pickle_step = {
                    'text': step_text,
                    'arguments': arguments,
                    'locations': [
                        _pickle_location(values['location'], path),
                        _pickle_step_location(scenario_outline_step, path)
                    ]
                }
                steps.append(_pickle_step)

            pickle = {
                'name': u'{0}: {1}'.format(
                    keyword,
                    _interpolate(
                        scenario_outline['name'],
                        variable_cells,
                        value_cells)
                    ),
                'steps': steps,
                'tags': _pickle_tags(tags, path),
                'locations': [
                    _pickle_location(values['location'], path),
                    _pickle_location(scenario_outline['location'], path)
                ]
            }
            pickles.append(pickle)


def _create_pickle_arguments(argument, variables, values, path):
    result = []

    if not argument:
        return result

    if argument['type'] is 'DataTable':
        table = {'rows': []}
        for row in argument['rows']:
            cells = [
                {
                    'location': _pickle_location(cell['location'], path),
                    'value': _interpolate(cell['value'], variables, values)
                } for cell in row['cells']
            ]
            table['rows'].append({'cells': cells})
        result.append(table)

    elif argument['type'] is 'DocString':
        docstring = {
            'location': _pickle_location(argument['location'], path),
            'content': _interpolate(argument['content'], variables, values)
        }
        result.append(docstring)

    else:
        raise Exception('Internal error')

    return result


def _interpolate(name, variable_cells, value_cells):
    for n, variable_cell in enumerate(variable_cells):
        value_cell = value_cells[n]
        name = re.sub(
            u'<{0[value]}>'.format(variable_cell),
            value_cell['value'],
            name
            )
    return name


def _get_background_steps(feature, path):
    if 'background' in feature:
        return [_pickle_step(step, path)
                for step in feature['background']['steps']]
    return []


def _pickle_step(step, path):
    return {
        'text': step['text'],
        'arguments': _create_pickle_arguments(
            step.get('argument'),
            [],
            [],
            path),
        'locations': [_pickle_step_location(step, path)]
    }


def _pickle_step_location(step, path):
    return {
        'path': path,
        'line': step['location']['line'],
        'column': step['location']['column'] + count_symbols(step.get('keyword', 0))
    }


def _pickle_location(location, path):
    return {
        'path': path,
        'line': location['line'],
        'column': location['column']
    }


def _pickle_tags(tags, path):
    return [_pickle_tag(tag, path) for tag in tags]


def _pickle_tag(tag, path):
    return {
        'name': tag['name'],
        'location': _pickle_location(tag['location'], path)
    }
